### PR TITLE
General improvements

### PR DESCRIPTION
**Pull request by nbianca**
Hello,

This patch is quite big, but there should not be anything complicated in particular.

* One of the most important changes is that the `insert-block` app event is used, which inserts the canned reply at the current position of the cursor, not at the end.

* Another change is that there is some basic support for variables that is currently _experimental_ and replaces `#{var_name}` with the value of variable `var_name`. Supported variables are: `user`, `op`, `last`, `replyto`, `replyto_or_last`.

* There are also plenty of changes that should improve the quality of the code overall (style changes, rendundancies, etc).

### CODE REVIEW

File: assets/javascripts/discourse/components/canned-reply.js.es6 [hunk 1]
```diff
@@ -15,18 +14,35 @@ export default Ember.Component.extend({
...
-          this.get("reply.content")
-        );
+      let replyTitle = this.get("reply.title");
+      let replyContent = this.get("reply.content");
 
-        if (!composer.model.get("title")) {
-          composer.model.set("title", this.get("reply.title"));
+      // Replace variables with values.
+      if (model) {
+        const vars = {
```
Comment (ZogStriP) [inline assets/javascripts/discourse/components/canned-reply.js.es6]: Where are these variables coming from/being used?

File: assets/javascripts/discourse/components/canned-reply.js.es6 [hunk 2]
```diff
@@ -17,9 +17,11 @@ export default Ember.Component.extend({
       const composer = Discourse.__container__.lookup("controller:composer");
 
       if (composer.model) {
-        const newReply =
-          composer.model.get("reply") + this.get("reply.content");
-        composer.model.set("reply", newReply);
+        this.appEvents.trigger(
```
Comment (ZogStriP) [inline assets/javascripts/discourse/components/canned-reply.js.es6]: This should be tested.

File: assets/javascripts/discourse/components/canned-reply.js.es6 [hunk 3]
```diff
@@ -6,11 +6,6 @@ import computed from "ember-addons/ember-computed-decorators";
 export default Ember.Component.extend({
   isOpen: false,
 
-  @computed("isOpen")
```
Comment (ZogStriP) [inline assets/javascripts/discourse/components/canned-reply.js.es6]: Why was this removed?

**Comment by ZogStriP** [comment]
We need some tests to ensure we don't regress ;)

File: .travis.yml [hunk 4]
```diff
@@ -2,45 +2,57 @@ language: ruby
 
 env:
   global:
+    - TRAVIS_NODE_VERSION="10"
```
[the remark this answers is not in the document]
Comment (nbianca) [reply .travis.yml]: I changed this file to keep it more or less in sync with [`.travis.yml` from discourse repo](https://github.com/discourse/discourse/blob/master/.travis.yml).

[the remark this answers is not in the document]
Comment (gschlager) [reply .travis.yml]: Why not test if https://meta.discourse.org/t/setting-up-plugin-continuous-integration-tests-on-travis-ci/59612 works as advertised. 😉

File: assets/javascripts/discourse/components/canned-reply.js.es6 (hunk 1 above)
Comment (nbianca) [reply assets/javascripts/discourse/components/canned-reply.js.es6]: This is a new feature that has been requested which allows replacing variables like #{user} with the username of current user.

The variables are currently "hardcoded", but it is a feature that has been requested for long time and the implementation was not difficult.

File: assets/javascripts/discourse/components/canned-reply.js.es6 (hunk 2 above)
Comment (nbianca) [reply assets/javascripts/discourse/components/canned-reply.js.es6]: Tests are coming soon. :dancer:

File: assets/javascripts/discourse/components/canned-reply.js.es6 (hunk 1 above)
Comment (ZogStriP) [reply assets/javascripts/discourse/components/canned-reply.js.es6]: Can you link to the conversion(s) on meta?

File: assets/javascripts/discourse/components/canned-reply.js.es6 (hunk 3 above)
Comment (nbianca) [reply assets/javascripts/discourse/components/canned-reply.js.es6]: `showContentIcon` was not used anywhere. It was removed in 4eb75be5e885d01a31cb2237db124d437c25cced.

File: .travis.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidtaylorhq) [reply .travis.yml]: It should do ;) - I set it up at https://travis-ci.org/discourse/discourse-chat-integration/

@nbianca let me know if you would like any help getting it working

File: assets/javascripts/lib/apply-reply.js.es6 [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+export default function(replyId, replyTitle, replyContent, model) {
+  // Replace variables with values.
+  if (model) {
+    const vars = {
+      user: model.get("user.username"),
+      op: model.get("topic.posters")
+        ? model.get("topic.posters")[0].user.get("username")
+        : "",
+      replyto: model.get("post.reply_to_user.username"),
```
Comment (ZogStriP) [inline assets/javascripts/lib/apply-reply.js.es6]: I think this one should be name `replyTo`.

File: assets/javascripts/lib/apply-reply.js.es6 [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { ajax } from "discourse/lib/ajax";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+
+export default function(replyId, replyTitle, replyContent, model) {
+  // Replace variables with values.
+  if (model) {
+    const vars = {
+      user: model.get("user.username"),
+      op: model.get("topic.posters")
```
Comment (ZogStriP) [inline assets/javascripts/lib/apply-reply.js.es6]: `original_poster` is better

File: assets/javascripts/lib/apply-reply.js.es6 [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        ? model.get("topic.posters")[0].user.get("username")
+        : "",
+      replyto: model.get("post.reply_to_user.username"),
+      last: model.get("topic.last_poster_username"),
```
Comment (ZogStriP) [inline assets/javascripts/lib/apply-reply.js.es6]: This one should be `last_user`.

File: assets/javascripts/lib/apply-reply.js.es6 [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      replyto_or_last:
```
Comment (ZogStriP) [inline assets/javascripts/lib/apply-reply.js.es6]: Should be `reply_to_or_last`

File: assets/javascripts/lib/apply-reply.js.es6 [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+      reply_to: model.get("post.reply_to_user.username"),
+      last_user: model.get("topic.last_poster_username"),
+      reply_to_or_last:
+        model.get("post.reply_to_user.username") ||
+        model.get("topic.last_poster_username")
+    };
+
+    for (var key in vars) {
+      if (vars[key]) {
+        replyTitle = replyTitle.replace("#{" + key + "}", vars[key]);
```
Comment (ZogStriP) [inline assets/javascripts/lib/apply-reply.js.es6]: Let's use `%{variable_name}` instead of `#{variable_name}` so that it's the same syntax as translated text.

**Comment by worldismine** [comment]
Ever since this change was implemented I am getting an odd and hard to reproduce problem with adding canned responses. Sometimes they work, sometimes they don't. When they don't, I get the following error in the console:

`Uncaught TypeError: i.get(...)[0].user.get is not a function`

It points to this snipplet:

` original_poster: model.get("topic.posters") ? model.get("topic.posters")[0].user.get("username") : "",`

I've tried on 2.2.2 with stock theme and no custom plugins. This behavior only happens in PM's (at random) where there are more than 2 people involved. I've tried for weeks to replicate this, but I can't. Sometimes it happens every time, sometimes it won't for hours.
